src/algebra/char_p/basic.lean

```diff
@@ -428,3 +428,21 @@ begin
 end
 
 end
+
+section prod
+
+variables (S : Type v) [semiring R] [semiring S] (p q : ℕ) [char_p R p]
+
+/-- The characteristic of the product of rings is the least common multiple of the
+characteristics of the two rings. -/
+instance [char_p S q] : char_p (R × S) (nat.lcm p q) :=
+{ cast_eq_zero_iff :=
+    by simp [prod.ext_iff, char_p.cast_eq_zero_iff R p,
+      char_p.cast_eq_zero_iff S q, nat.lcm_dvd_iff] }
```

Comment:
Not sure if I like this instance: won't you get a lot if problems where you want e.g. `char_P (R × S) 6` but you get `char_p (R × S) (nat.lcm 2 3)`? Instead, I'm inclined to have a `theorem [char_p S q] (h : nat.lcm p q = r)  : char_p (R × S) r` that can be instantiated in specific cases like `prod.char_p` below.

Reply:
As i comment above, instance search doesn't work like that, at least not when `p` is not an `out_param`

Reply:
(Link to the comment: https://github.com/leanprover-community/mathlib/pull/8356#discussion_r672184812)

I see, if it's not an `out_param` then this should be robust enough. Still not completely convinced that it will be convenient (e.g. `example : ((6 : ℕ) : zmod 2 × zmod 3) = 0 := char_p.cast_eq_zero _ (nat.lcm 2 3)` doesn't work), but at least it won't hurt.